### ISSUE TITLE
Handling without control

### ISSUE THREAD

**Issue by phischu** [effekt-lang/effekt#252]
It would be great if objects could be passed implicitly without introducing an effect handler.

```
interface State {
  def get(): Int
  def set(x: Int): Unit
}

def myComputation(): Int / State = do get()

def main() = {
  var s = 0;
  with new State {
    def get() = s
    def set(x: Int) = s = x
  };
  myComputation()
}
```

The typing rule for `with` should be like

```
G |- b : F    G |- s : 𝜏 / ε
-----------------
G |- with b; s : 𝜏 / (ε - F)
```

**Comment by b-studios** [comment]
Yes, this is on our map for quite some time. I have thought a bit about it: most of it should be fairly straight forward. There are a few things that need to be sorted out:

- what syntactic category (in source) is `with`? Can be a `Stmt`, or a `Def`, or a `Term` -- all would make sense. Maybe `Def` is the most appropriate one.
- much more important: how does this interact with capture tracking (System C). I would assume the outer context captures whatever b and s capture, but the inferred capability would probably need to be transparent. This is a problem, since in Typer associates effect types with tracked block parameters: https://github.com/effekt-lang/effekt/blob/13af16fc1215814905cba9d0839af862d8bd00a1/effekt/shared/src/main/scala/effekt/Typer.scala#L1289

**Comment by kyay10** [comment]
Since `with` exists already, all you'd really need is some way to turn an explicit capability into an implicit one, like #406 suggests (and also #67 block type polymorphism to make it simpler)

If we want usage like:
```effekt
{
  with given{myEffect}
  do someMyEffectOp()
}
```

Currently you can do:
```effekt
def given[R]{cap: MyEffect} {block: => R / {MyEffect}} = try block() with MyEffect {
  // delegate all methods tail-resumptively
}
```
you'd have to write it for every type though.
With #406, you can do:
```effekt
def given[R]{cap: MyEffect} {block: => R / {MyEffect}} = try block() with cap // magic syntax that doesn't exist yet
```
but you'd still have to write it for each type
and with #67 you can finally do:
```effekt
def given[E, R]{cap: E} {block: => R / {E}} = try block() with cap
```
which can just go in the prelude for everyone to use!

**Comment by phischu** [comment]
Syntax-wise I believe we can use `with new MyEffect { ... }; ...`.

Last time we checked there was some difficulty because the captures of implicit things must always be opaque but here they should be transparent.
